### PR TITLE
Adding IREE parameter archive format and tooling support.

### PR DESCRIPTION
**Pull request by benvanik**
The new format allows us to store parameters for both inference and training observing the requirements for both efficient CPU and GPU/accelerator execution. We can also support additional storage types such as splats allowing for stripped parameter files that work with programs compiled assuming real parameters. The format has a provision for referencing external file ranges but support for reading such files is TBD.

The format can be used like gguf/safetensors and is supported by the tooling in the same way. Additionally a new `iree-convert-parameters` tool is added to convert any format supported for loading (gguf, safetensors, and irpa itself) into irpa files with some control over which parameters are included, renaming of parameters, and stripping parameters and replacing them with splat values. This should make it easy to take any gguf/safetensors file and quickly create stripped variants for easy reproducers/CI benchmarking without needing to ship around the original files. The `iree-create-parameters` tool can be used to create empty archives that are ready for initialization from a program that mutates them or to create parameter archives with named parameters when there is no source gguf/safetensors file.

All of this is still using memory-mapped files; this limits our parameter file sizes on 32-bit systems but I suspect no one is going to run this tool for large models on 32-bit systems. In the future we can make the conversion tool use the HAL and schedule out optimized file I/O. For now we just copy parameters via a normal read/write loop and it's fastish-enough (pretty much I/O bound, with less optimal reads because of memory mapping). For me with a cold cache it takes ~1min to rewrite a 25GB file and 25sec with a hot cache.

This initial commit has the IRPA builder using the new iree_io_stream_t but switching all format parsers to use it is deferred to future changes.

Progress on #15521.

### CODE REVIEW

**Comment by stellaraccident** [comment]
Ping me when this is ready for a full review.

**Comment by benvanik** [comment]
This is ready for review @stellaraccident - still need to add some tests but all functionality is there and should be working (can try out iree-create-parameters/iree-convert-parameters).

I don't understand the pytype error - any ideas? https://github.com/openxla/iree/actions/runs/7012119296/job/19076020636?pr=15670
```
Running pycheck against 'main'
Checking .py files outside of integrations/
ninja: Entering directory `.pytype'
ninja: error: build.ninja:1[6](https://github.com/openxla/iree/actions/runs/7012119296/job/19076020636?pr=15670#step:6:7): multiple rules generate /home/runner/work/iree/iree/.pytype/pyi/__main__.pyi

Computing dependencies
Analyzing 4 sources with 0 local dependencies
Leaving directory '.pytype'

Checking .py files in integrations/.*
No files to check.

One or more pytype checks failed.
You can view these errors locally by running
  ./build_tools/pytype/check_diff.sh main
```

**Comment by ScottTodd** [comment]
> I don't understand the pytype error - any ideas?

Likely safe to ignore... IIRC that happens whenever a `__main__.py` is touched and we haven't taken the time to find a workaround/fix.

**Comment by stellaraccident** [comment]
Just getting to this and need to sign off for the day. But I got my task list down to zero... So tomorrow :)

**Comment by benvanik** [comment]
Basic IRPA parser test is in - gotta get this landed to unblock people then will loop back and rewrite all the parsers to use `iree_io_stream_t` and add round-trip testing and other things for the IRPA files. Immediate next steps are to start working through some of the bugs/scaling issues in #15521 / #15522 / #15674.

File: runtime/src/iree/schemas/parameter_archive.h [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+#include <stdint.h>
+
+//===----------------------------------------------------------------------===//
+// IREE Parameter Archive
+//===----------------------------------------------------------------------===//
+//
+// Each parameter archive stores zero or more parameters that are referenced by
+// a string name. A single file may contain multiple archives linked together.
+// Archives are designed for optimized concatenation (extend a file and link
+// each archive header), appended (link new archive header), erasing (change
```
Comment (ScottTodd) [inline runtime/src/iree/schemas/parameter_archive.h]: grammar nit?
```suggestion
// each archive header), appending (link new archive header), erasing (change
```

File: runtime/src/iree/schemas/parameter_archive.h [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+// entry to skip), and replacing (existing skip and then a new archive header).
+//
+// During mutation the archives may get spread out over the file and hurt
+// startup performance. Repacking archive files with `iree-repack-parameters`
+// will combine all archive headers at the head of the file to avoid needing to
+// scan large files to index their contents.
```
Comment (ScottTodd) [inline runtime/src/iree/schemas/parameter_archive.h]: Does `iree-repack-parameters` exist in this PR?

File: runtime/src/iree/io/formats/safetensors/safetensors_parser.h [hunk 3]
```diff
@@ -4,8 +4,8 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-#ifndef IREE_IO_FORMATS_SAFETENSORS_SAFETENSORS_FORMAT_H_
-#define IREE_IO_FORMATS_SAFETENSORS_SAFETENSORS_FORMAT_H_
+#ifndef IREE_IO_FORMATS_SAFETENSORS_safetensors_parser_H_
+#define IREE_IO_FORMATS_SAFETENSORS_safetensors_parser_H_
```
Comment (ScottTodd) [inline runtime/src/iree/io/formats/safetensors/safetensors_parser.h]: revert find/replace changes, keep the caps in the header guard

File: runtime/src/iree/io/formats/irpa/irpa_parser.c [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+          {
+              .splat =
+                  {
+                      // .pattern = splat_entry->pattern,  below
+                      .pattern_length = splat_entry->pattern_length,
+                  },
+          },
+  };
+  memcpy(entry.storage.splat.pattern, splat_entry->pattern,
+         entry.storage.splat.pattern_length);
```
Comment (ScottTodd) [inline runtime/src/iree/io/formats/irpa/irpa_parser.c]: I'm a bit confused by the comment here. Seems like a struct member is uninitialized here? Maybe set to NULL/0 explicitly?

File: runtime/src/iree/io/formats/irpa/irpa_builder.c [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+      .storage =
+          {
+              .splat =
+                  {
+                      // .pattern = ... set below ...,
+                      .pattern_length = pattern_length,
+                  },
+          },
+  };
+  memcpy(entry.storage.splat.pattern, pattern, pattern_length);
```
Comment (ScottTodd) [inline runtime/src/iree/io/formats/irpa/irpa_builder.c]: Same here. Set to NULL/0 explicitly instead of having just a comment?

**Review by ScottTodd — APPROVED** [review]

File: runtime/bindings/python/CMakeLists.txt [hunk 6]
```diff
@@ -140,6 +141,8 @@ iree_py_library(
     "iree/_runtime/scripts/iree_benchmark_trace/__main__.py"
     "iree/_runtime/scripts/iree_benchmark_module/__main__.py"
     "iree/_runtime/scripts/iree_cpuinfo/__main__.py"
+    "iree/_runtime/scripts/iree_convert_parameters/__main__.py"
```
Comment (stellaraccident) [inline runtime/bindings/python/CMakeLists.txt:144]: Note to self: I'm soon going to need to busybox these like on the compiler side. I think we've got like 6 copies of the runtime in our Python distro now (7.9MiB/4.2Mib on Linux/Windows). The only way that has worked this long is that it is pretty binary efficient as it is...

File: runtime/src/iree/io/formats/irpa/irpa_builder.c [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_add_splat_entry(
+    iree_io_parameter_archive_builder_t* builder, iree_string_view_t name,
+    iree_const_byte_span_t metadata, const void* pattern,
+    uint8_t pattern_length, iree_io_physical_size_t data_length) {
```
Comment (stellaraccident) [inline runtime/src/iree/io/formats/irpa/irpa_builder.c:223]: I think you need a check that pattern_length does not exceed the size of where it is being written.

File: runtime/src/iree/io/formats/irpa/irpa_builder.h [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+// appended to the index referencing the given |file_handle|.
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_write(
+    const iree_io_parameter_archive_builder_t* builder,
+    iree_io_file_handle_t* file_handle, iree_io_physical_offset_t file_offset,
+    iree_io_stream_t* stream, iree_io_parameter_index_t* target_index);
+
+// Adds a new splat entry to |builder|.
+// Splat entries have no physical storage and exist only in the header.
+// |pattern| and |metadata| (if provided) are copied prior to returning.
+IREE_API_EXPORT iree_status_t iree_io_parameter_archive_builder_add_splat_entry(
```
Comment (stellaraccident) [inline runtime/src/iree/io/formats/irpa/irpa_builder.h:77]: Maybe a note on maximum splat pattern size? I've mis-assumed this before and it cost me time...

File: runtime/src/iree/io/stream.c [hunk 9]
```diff
@@ -8,7 +8,7 @@
 
 // This is arbitrary - we should allow dynamic block sizes and such but keeping
 // this small only requires a reasonable fixed size stack alloc.
-#define IREE_IO_STREAM_COPY_BLOCK_SIZE (16 * 1024)
+#define IREE_IO_STREAM_COPY_BLOCK_SIZE (64 * 1024)
```
Comment (stellaraccident) [inline runtime/src/iree/io/stream.c]: Just for my own edification, which arbitrary thing did you find that made you want this bigger?

**Review by stellaraccident — APPROVED** [review]
Couple of small things plus scott's comments. Otherwise, lgtm.

File: runtime/src/iree/io/stream.c (hunk 9 above)
Comment (benvanik) [reply runtime/src/iree/io/stream.c]: Was just benchmarking 25GB parameter file rewrites - 32k is about the same as 64k for me so shrinking to that for now. There's some smarter things we can do if this ever becomes a bottleneck.

File: runtime/bindings/python/CMakeLists.txt (hunk 6 above)
Comment (benvanik) [reply runtime/bindings/python/CMakeLists.txt:144]: Yeah, you could at least combine a few of these (like all the parameter tools) that have the same deps. Things like iree-benchmark-module/iree-run-module/iree-check-module have different primary deps though (gtest/benchmark complicating things).

File: runtime/src/iree/io/formats/irpa/irpa_parser.c (hunk 4 above)
Comment (benvanik) [reply runtime/src/iree/io/formats/irpa/irpa_parser.c]: added zero set, but the memcpy is needed - the input pattern pointer only has pattern_length valid bytes and we can't statically initialize it inline

**Pull request merged**